plugins/aws/mfa_provisioner.go

```diff
@@ -41,39 +48,59 @@ func (p STSProvisioner) Provision(ctx context.Context, in sdk.ProvisionInput, ou
 		region = os.Getenv("AWS_DEFAULT_REGION")
 	}
 	if len(region) == 0 {
-		out.AddError(errors.New("region is required for the AWS Shell Plugin MFA workflow: set 'default region' in 1Password or set the 'AWS_DEFAULT_REGION' environment variable yourself"))
+		out.AddError(errors.New("region is required for the AWS Shell Plugin MFA or Assume Role workflows: set 'default region' in 1Password or set the 'AWS_DEFAULT_REGION' environment variable yourself"))
 		return
 	}
 	config.Region = region
 
 	stsProvider := sts.NewFromConfig(*config)
+
+	if useCache := tryUsingCachedCredentials(mfaCacheKey, in, out); useCache {
+		return
+	}
 	input := &sts.GetSessionTokenInput{
-		DurationSeconds: aws.Int32(900), // minimum expiration time - 15 minutes
+		DurationSeconds: aws.Int32(int32(p.DurationSeconds.Seconds())),
 		SerialNumber:    aws.String(p.MFASerial),
 		TokenCode:       aws.String(p.TOTPCode),
 	}
 
-	result, err := stsProvider.GetSessionToken(ctx, input)
+	resp, err := stsProvider.GetSessionToken(ctx, input)
 	if err != nil {
 		out.AddError(err)
 		return
 	}
-
-	out.AddEnvVar("AWS_ACCESS_KEY_ID", *result.Credentials.AccessKeyId)
-	out.AddEnvVar("AWS_SECRET_ACCESS_KEY", *result.Credentials.SecretAccessKey)
-	out.AddEnvVar("AWS_SESSION_TOKEN", *result.Credentials.SessionToken)
-	out.AddEnvVar("AWS_DEFAULT_REGION", region)
-
-	err = out.Cache.Put("sts", *result.Credentials, *result.Credentials.Expiration)
+	out.AddEnvVar("AWS_ACCESS_KEY_ID", *resp.Credentials.AccessKeyId)
+	out.AddEnvVar("AWS_SECRET_ACCESS_KEY", *resp.Credentials.SecretAccessKey)
+	out.AddEnvVar("AWS_SESSION_TOKEN", *resp.Credentials.SessionToken)
+	err = out.Cache.Put(mfaCacheKey, *resp.Credentials, *resp.Credentials.Expiration)
 	if err != nil {
 		out.AddError(fmt.Errorf("failed to serialize aws sts credentials: %w", err))
 	}
+	out.AddEnvVar("AWS_DEFAULT_REGION", region)
+
+}
+
+func tryUsingCachedCredentials(cacheKey string, in sdk.ProvisionInput, out *sdk.ProvisionOutput) bool {
+	var cached types.Credentials
+	if ok := in.Cache.Get(cacheKey, &cached); ok {
+
+		out.AddEnvVar("AWS_ACCESS_KEY_ID", *cached.AccessKeyId)
+		out.AddEnvVar("AWS_SECRET_ACCESS_KEY", *cached.SecretAccessKey)
+		out.AddEnvVar("AWS_SESSION_TOKEN", *cached.SessionToken)
+
+		if region, ok := in.ItemFields[fieldname.DefaultRegion]; ok {
+			out.AddEnvVar("AWS_DEFAULT_REGION", region)
+		}
+
+		return true
+	}
+	return false
 }
 
-func (p STSProvisioner) Deprovision(ctx context.Context, in sdk.DeprovisionInput, out *sdk.DeprovisionOutput) {
-	// Nothing to do here: environment variables get wiped automatically when the process exits.
+func (p mfaProvisioner) Deprovision(ctx context.Context, in sdk.DeprovisionInput, out *sdk.DeprovisionOutput) {
+	// nothing to do here: 1Password CLI removes env vars
```

Comment:
This sounds to me like 1Password CLI is actively pruning the envvars. This isn't what's happening, instead, they are wiped by the operating system when the process exits. For that reason, I liked the previous version of this comment better.

Reply:
Alright, makes sense, done!